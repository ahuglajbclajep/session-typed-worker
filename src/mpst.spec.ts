import type { Select, Offer, Close, Merge } from "./mpst";

declare const ms0: Merge<
  | Select<"A", { a: (v: string) => Close }>
  | Select<"A", { a: (v: string) => Close }>
>;
// @dts-jest:pass:snap Select<"A", { a: (v: string) => Close }>
ms0;

declare const ms1: Merge<
  | Select<"A", { a: (v: string) => Close }>
  | Select<"B", { a: (v: string) => Close }>
>;
// @dts-jest:pass:snap MPSTError < "Select: destination role conflict", "A" | "B" >
ms1;

declare const ms2: Merge<
  | Select<"A", { a: (v: string) => Close }>
  | Select<"A", { b: (v: string) => Close }>
>;
// @dts-jest:pass:snap Select<"A", MPSTError<"SelectBranch: labels differ", "a" | "b">>
ms2;

declare const ms3: Merge<
  | Select<"A", { a: (v: string) => Close }>
  | Select<"A", { a: (v: number) => Close }>
>;
// @dts-jest:pass:snap -> Select<"A", { a: MPSTError<"SelectBranch: values differ", string | number>; }>
ms3;

declare const mo0: Merge<
  Offer<"A", ["a", [string, Close]]> | Offer<"A", ["a", [string, Close]]>
>;
// @dts-jest:pass:snap Offer<"A", ["a", [string, Close]]>
mo0;

declare const mo1: Merge<
  Offer<"A", ["a", [string, Close]]> | Offer<"B", ["a", [string, Close]]>
>;
// @dts-jest:pass:snap MPSTError<"Offer: destination role conflict", "A" | "B">
mo1;

declare const mo2: Merge<
  Offer<"A", ["a", [string, Close]]> | Offer<"A", ["b", [string, Close]]>
>;
// @dts-jest:pass:snap Offer<"A", ["a", [string, Close]] | ["b", [string, Close]]>
mo2;

declare const mo3: Merge<
  Offer<"A", ["a", [string, Close]]> | Offer<"A", ["a", [number, Close]]>
>;
// @dts-jest:pass:snap Offer<"A", MPSTError<"OfferBranch: values differ", string | number>>
mo3;

declare const mc0: Merge<Close | Close>;
// @dts-jest:pass:snap Close
mc0;

declare const me0: Merge<
  Select<"A", { a: (v: string) => Close }> | Offer<"A", ["a", [string, Close]]>
>;
// @dts-jest:pass:snap MPSTError<"Merge: local type conflict", "select" | "offer">
me0;
