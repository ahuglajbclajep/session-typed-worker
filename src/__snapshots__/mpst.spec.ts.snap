// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Close (type) should match snapshot 1`] = `"Close"`;

exports[`G0["B"] (type) should match snapshot 1`] = `"Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Close; }>]] | [\\"b\\", [{}, Select<\\"C\\", { b1: (v: {}) => Close; }>]]>"`;

exports[`G1["A"] (type) should match snapshot 1`] = `"Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Select<\\"B\\", { a: (v: {}) => any; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>]]>; b: (v: {}) => Close; }>"`;

exports[`G1["B"] (type) should match snapshot 1`] = `"Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => any; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>"`;

exports[`G1["B"] (type) should match snapshot 2`] = `"Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => Offer<\\"A\\", [\\"a\\", [{}, Select<\\"C\\", { a1: (v: string) => any; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>; }>]] | [\\"b\\", [{}, Select<\\"C\\", { a1: (v: {}) => Close; }>]]>"`;

exports[`MPSTError < "Select: destination role conflict", "A" | "B" > (type) should match snapshot 1`] = `"MPSTError<\\"Select: destination role conflict\\", \\"A\\" | \\"B\\">"`;

exports[`MPSTError<"Merge: local type conflict", "offer" | "select"> (type) should match snapshot 1`] = `"MPSTError<\\"Merge: local type conflict\\", \\"select\\" | \\"offer\\">"`;

exports[`MPSTError<"Merge: local type conflict", "offer"> (type) should match snapshot 1`] = `"MPSTError<\\"Merge: local type conflict\\", \\"offer\\">"`;

exports[`MPSTError<"Merge: local type conflict", "select" | "offer"> (type) should match snapshot 1`] = `"MPSTError<\\"Merge: local type conflict\\", \\"select\\" | \\"offer\\">"`;

exports[`MPSTError<"Offer: destination role conflict", "A" | "B"> (type) should match snapshot 1`] = `"MPSTError<\\"Offer: destination role conflict\\", \\"A\\" | \\"B\\">"`;

exports[`Offer<"A", ["a", [string, Close]] | ["b", [string, Close]]> (type) should match snapshot 1`] = `"Offer<\\"A\\", [\\"a\\", [string, Close]] | [\\"b\\", [string, Close]]>"`;

exports[`Offer<"A", ["a", [string, Close]]> (type) should match snapshot 1`] = `"Offer<\\"A\\", [\\"a\\", [string, Close]]>"`;

exports[`Offer<"A", MPSTError<"OfferBranch: values differ", string | number>> (type) should match snapshot 1`] = `"Offer<\\"A\\", MPSTError<\\"OfferBranch: values differ\\", string | number>>"`;

exports[`Offer<"B", ["a1", [string, Select<"A", { a2: (v: boolean) => Close }>]] | ["b1", [{}, Close]]> (type) should match snapshot 1`] = `"Offer<\\"B\\", [\\"a1\\", [string, Select<\\"A\\", { a2: (v: boolean) => Close; }>]] | [\\"b1\\", [{}, Close]]>"`;

exports[`Offer<"B", MPSTError<"OfferBranch: values differ", string | {}>> (type) should match snapshot 1`] = `"Offer<\\"B\\", MPSTError<\\"OfferBranch: values differ\\", string | {}>>"`;

exports[`Select<"A", { a: (v: string) => Close }> (type) should match snapshot 1`] = `"Select<\\"A\\", { a: (v: string) => Close; }>"`;

exports[`Select<"A", { a: MPSTError<"SelectBranch: values differ", string | number>; }> (type) should match snapshot 1`] = `"Select<\\"A\\", { a: MPSTError<\\"SelectBranch: values differ\\", string | number>; }>"`;

exports[`Select<"A", MPSTError<"SelectBranch: labels differ", "a" | "b">> (type) should match snapshot 1`] = `"Select<\\"A\\", MPSTError<\\"SelectBranch: labels differ\\", \\"a\\" | \\"b\\">>"`;

exports[`Select<"B", { a: (v: {}) => Offer<"C", ["a2", [boolean, Close]]>; b: (v: {}) => Close }> (type) should match snapshot 1`] = `"Select<\\"B\\", { a: (v: {}) => Offer<\\"C\\", [\\"a2\\", [boolean, Close]]>; b: (v: {}) => Close; }>"`;
